specs/request-handler-plugins.md

```diff
@@ -0,0 +1,116 @@
+- `configSection` is a string for which there must be a matching object property at the root of the configuration file. If a handler does not provide any configuration options then an empty object must be supplied.
+- `handlerPath` a string which is the relative path to the assembly containing the class definition for the handler which will be loaded using reflection.
+- `name` a string identifier for the handler class, allows developer to ship multiple handlers in a single assembly.
+- `urlsToWatch` is an array of strings defining the urls for which this hander will watch, this behavior is defined in the [Multi URL support spec](./multi-url-support.md).
+
+```json
+  "handlers": [
+    {
+        "configSection": "msGraphProxy",
+        "handlerPath": "GraphHandler\\msgraph-proxy-handler.dll",
+        "name": "GraphHandler",
+        "urlsToWatch": [
+            "https://graph.microsoft.com/v1.0/*",
+            "https://graph.microsoft.com/beta/*"
+        ],
+    }
+  ],
+  "msGraphProxy": {
+    // config options specific to the msGraphProxy
+  }
+```
+
+Handlers will executed in the order that they are listed in the handlers configuration array. At least one handlers must be present in the configuration, otherwise the proxy will exit with an error informing the user that their configuration is invalid and contains no handlers.
+
+## Handlers
+
+Handler classes must inherit from the provider abstract base class `ProxyHandler`, proposed implementation listed below for reference, and must provide an implementation for the `ExecuteInternal` method.
+Handlers can set the response which will be provided by the proxy using the `GenericResponse` method of the provided `SessionEventArgs` object, when a handler sets the response it should also set the `IsComplete` property of the provided `ResponseState` to prevent subsequently executed handler from modifying the response.
+
+```cs
+using System.CommandLine;
+using System.CommandLine.Invocation;
+using System.Text.RegularExpressions;
+using Microsoft.Extensions.Configuration;
+using Titanium.Web.Proxy.EventArguments;
+
+public abstract class ProxyHandler
+{
+    private readonly IEnumerable<Regex> _urlsToWatch;
+    private readonly IConfigurationSection _config;
+
+    public ProxyHandler(IEnumerable<Regex> urlsToWatch, IConfigurationSection config)
+    {
+        this._urlsToWatch = urlsToWatch ?? throw new ArgumentNullException(nameof(urlsToWatch));
+        if (!urlsToWatch.Any())
+            throw new ArgumentException($"{nameof(urlsToWatch)} cannot be empty", nameof(urlsToWatch));
+        this._config = config;
+    }
+
+    /**
+     * Use this method to add any command line options to the proxy
+     */
+    public abstract void AddCommandLineOptions(RootCommand command);
+
+    /**
+     * Read options from command line args and update internal configuration
+     */
+    public abstract int UpdateConfiguration(InvocationContext context);
```

Comment:
What is the `int` value that we're returning here?

Reply:
A return code for the proxy, non-zero on error to trigger the proxy to exit.

Reply:
This is irrelevant now, under the event-based model implementers would need to throw unhandled exceptions to trigger the proxy to exit.